test/unit/math/prim/prob/skew_double_exponential_test.cpp

```diff
@@ -0,0 +1,66 @@
+#include <stan/math/prim.hpp>
+#include <test/unit/math/prim/prob/vector_rng_test_helper.hpp>
+#include <test/unit/math/prim/prob/util.hpp>
+#include <gtest/gtest.h>
+#include <boost/math/distributions.hpp>
+#include <boost/random/mersenne_twister.hpp>
+#include <limits>
+#include <vector>
+
+class SkewDoubleExponentialTestRig : public VectorRNGTestRig {
+ public:
+  SkewDoubleExponentialTestRig()
+      : VectorRNGTestRig(10000, 10, {-2.5, -1.7, -0.1, 0.1, 2.0},
+                         {-3, -2, -1, 0, 2, 6}, {}, {}, {0.1, 1.0, 2.5, 4.0},
+                         {1, 2, 3, 4}, {-2.7, -1.5, -0.5, 0.0}, {-3, -2, -1, 0},
+                         {0.5}, {0}, {-0.1, 1.1}, {}) {}
+
+  template <typename T1, typename T2, typename T3, typename T_rng>
+  auto generate_samples(const T1& mu, const T2& sigma, const T3& tau,
+                        T_rng& rng) const {
+    return stan::math::skew_double_exponential_rng(mu, sigma, tau, rng);
+  }
+};
+
+TEST(ProbDistributionsSkewedDoubleExponential, errorCheck) {
+  check_dist_throws_all_types(SkewDoubleExponentialTestRig());
+}
+
+TEST(ProbDistributionsSkewedDoubleExponential, error_check) {
+  boost::random::mt19937 rng;
+  EXPECT_NO_THROW(stan::math::skew_double_exponential_rng(10.0, 2.0, .1, rng));
+
+  EXPECT_THROW(stan::math::skew_double_exponential_rng(10.0, 2.0, -1.0, rng),
+               std::domain_error);
+  EXPECT_THROW(stan::math::skew_double_exponential_rng(
+                   10, 2, stan::math::positive_infinity(), rng),
+               std::domain_error);
+}
+
+/*
+ * We test the skew double exponential rng by setting tau=0.5 which recovers
+ * a conventional double exponential distribution from Boost
+ */
+TEST(ProbDistributionsSkewedDoubleExponential, chiSquareGoodnessFitTest) {
+  boost::random::mt19937 rng;
+  int N = 10000;
+  int K = stan::math::round(2 * std::pow(N, 0.4));
+
+  std::vector<double> samples;
+  for (int i = 0; i < N; ++i) {
+    samples.push_back(
+        stan::math::skew_double_exponential_rng(2.0, 1.0, 0.5, rng));
```

Comment:
We should test this with something slightly skew.

I think the way to generate reference random numbers is to do it by a mixture of two exponential distributions, like one positive, and one negative, if that makes sense.

I'm not sure why we generate the reference quantiles with rngs? We have the quantiles all here analytically. Maybe it's something to do with the chi-square test? Anyway, if we write two different ways to generate skew double exponential and they math, then we're probably good.

If you push something up I'll happily print out the quantiles and check them against our CDF function as well (I'm pretty sure that's right)

Reply:
Sorry @bbbales2, I'm not entirely following: 

> I think the way to generate reference random numbers is to do it by a mixture of two exponential distributions, like one positive, and one negative, if that makes sense.

How, would the two be parametrized? And this would go into the goodness of fit test?

> I'm not sure why we generate the reference quantiles with rngs? 

To be honest, I just copy and pasted it from the `skew_normal_test`. Could you give me a hint, what I should be replacing exactly? Sorry for the trouble!

Cheers,
Simon

Reply:
Added another comment below! Ignore the mixture of exponentials thing.